src/main/java/hudson/plugins/msbuild/MsBuildBuilder.java

```diff
@@ -29,44 +29,49 @@
 import hudson.tasks.Builder;
 import hudson.tools.ToolInstallation;
 import hudson.util.ArgumentListBuilder;
-import org.jenkinsci.Symbol;
-import org.kohsuke.stapler.DataBoundConstructor;
-
 import java.io.IOException;
+import java.nio.charset.Charset;
 import java.util.Arrays;
 import java.util.Map;
 import java.util.Set;
-import java.nio.charset.Charset;
+import jenkins.model.Jenkins;
+import jenkins.tasks.SimpleBuildStep;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
 
 /**
  * @author kyle.sweeney@valtech.com
  */
-public class MsBuildBuilder extends Builder {
+public class MsBuildBuilder extends Builder implements SimpleBuildStep {
```

Comment:
This is not really a good idea. A `SimpleBuildStep` should be used only for very short-lived commands: it cannot survive an agent or master restart. The actual build should be run with plain old

```groovy
powershell 'msbuild whatever'
```

If there is a compelling need to manage installations of `msbuild.exe` then you can use the `tool` step in advance; or if setting up critical environment variables is especially onerous you can define a `SimpleBuildWrapper` (or explicit block-scoped Pipeline `Step`) named something like `withMsbuild` to help automate that aspect while still leaving the actual tool invocation to the `powershell` (or `bat`) step.

Anyway these days it is best to run inside a defined container (or at least VM) environment so that there is no need for a special plugin at all. Plugins to manage and run tools are really only appropriate for legacy installations on manually-managed hardware.

Reply:
ok, that sounds like the right mindset, but how would you integrate the msbuild to a pipeline?

Reply:
Well, depending on details of your project and environment (and bearing in mind that I know nothing about how MSBuild works so I am assuming it is just roughly analogous to Linux’ `make`) it could just look like

```groovy
node('suitably-configured-windows') {
  checkout scm
  bat 'msbuild buildapp.csproj -t:HelloWorld' // Googled this
}
```

Reply:
thanks for the quick answers, but that doesn't "see" the entire problem.
msbuild comes in many shapes and forms:

1. msbuild.exe versions (according to the visual studio they came with)
2. various parameters to the msbuild.exe itself (after the project name to compile... like platform sdk, log verbosity, etc...)
3. different installation parameters (msbuild.exe file location on the specific machine)

sorry, I don't see how your suggestion start to handle all of these dimensions...

I hope I'm not missing something here...

Reply:
As I say, I am not familiar with details of MSBuild and cannot offer specific advice.

Reply:
We use MSBuild heavily with Jenkins Scripted and Declarative pipelines. The MSBuild plugin allows us to define the various tool locations (for Visual Studio 2019, 2017, etc.), including the ability to override it as needed on individual build nodes. Given this, we can then do:

```
script {
  def msbuild = tool name: 'MSBuild 2017 x64', type: 'hudson.plugins.msbuild.MsBuildInstallation'
  bat "${msbuild} SomeSolution.sln"
}
```

However, we've also dropped this into a global Shared Library so now we can just do this:

```
runMSBuild(version: 'MSBuild 2017 x64', project: 'SomeSolution.sln')
```

It works very well for us.

Reply:
thanks @medianick!

this is EXACTLY what I was looking for!

any chance you can share the groovy library here? (is it on github?)

Reply:
If you have a repo configured as a [Shared Library](https://jenkins.io/doc/book/pipeline/shared-libraries/), with `vars/runMSBuild.groovy` in it, just put something like this inside:

```
def call(String version, String project, String target = 'Build', String parameters = '') {
  def _installation = tool name: version, type: 'hudson.plugins.msbuild.MsBuildInstallation'

  bat "${_installation} ${project} /t:${target} ${parameters}"
}
```

Reply:
I'm sorry to say this solution has a flaw. when I tried using it in our environment, the only msbuild.exe path I could get is the one defined in the tool definition section of the jenkins server itself.

in our case, some of the slave nodes have msbuild installed in different locations. when using the "freestyle" project, each job execution got the right location of the msbuild.exe on the slave is was running on. when using your suggestion inside pipeline, I only get the msbuil path definition from the server.

do you have any suggestion?

Reply:
There is a node property for custom tool installation location, which the Pipeline `tool` step should honor when run inside an agent `node` block. Of course it would be simplest if you could just expect `msbuild.exe` to be in `%Path%` on all of your build machines.

Reply:
thanks for the quick answer, but we have more than one version of msbuild on our slaves (so having msbuild in the path won't do the trick).
you mentioned "node property for custom tool installation location" - can you please give a code snippet of how to use it? I don't understand exactly what you mean...

many thanks in advance :)

Reply:
This, under the Configure screen for each node:
![image](https://user-images.githubusercontent.com/329902/70178316-0dccd300-16aa-11ea-8c1c-5deb0f88a46e.png)

Reply:
For example code see https://github.com/jenkinsci/pipeline-examples/blob/master/jenkinsfile-examples/msbuild/Jenkinsfile, where "MSBuild" in '`${tool 'MSBuild'}`' would match one of the names posted by Medianick, so for example "`{tool '(MSBuild) MSBuild 2017'}`"

Reply:
do you know if the answer you provided suppose to work for a node defained using the [swarm](https://plugins.jenkins.io/swarm) plugin?

Reply:
OK, so now I solved the previous issue but I have a different question.
is there a way to get a list of all the installed msbuild tool definitions?

Reply:
You'd have to look at the tools available to Jenkins. 
global = http://*JenkinsURL*/configureTools
node specific = http://*JenkinsURL*/computer/*NodeName*/configure

Reply:
@DamonStamper - is there a pipeline api to do that?

Reply:
I've never needed to do that so I can't speak from experience. However a quick Google shows https://stackoverflow.com/a/35292719/7902967. If it can be done in the script console then you should be able to do it from a jenkinsfile. I'd suggest running the check in the agent/node block otherwise you'd only get globally defined tools.

Reply:
I know how to get the list of all the intalled plugins..
I think I didn't explained my self properly. 

what I'm looking for is a way to handle a case where a user is entering a wrong name of msbuild installation. I want to list all the available msbuild installations available on the specific node the user tried to execute the job on...

Reply:
Failing the build when the tool isn't found and then having the error message list out the available tools seems like the way to go. The other option would be to use something like ActiveChoices plugin to provide a list of choices and not let the user manually type something in. That plugin would only work if you are doing a freestyle job though.

Reply:
I'm glad I was able to explain my goal - you got it right! :)

the thing is, we are not using freestyle jobs anymore...

do you have any other suggestions?

Reply:
See the first sentence of my last reply.

Reply:
thanks for all the help. one last question:
is there a way to query the plugin using the following [api ](https://javadoc.jenkins.io/plugin/msbuild/hudson/plugins/msbuild/MsBuildInstallation.html#forNode-hudson.model.Node-hudson.model.TaskListener-) from a running pipeline? (the forNode method)